consensus/scp/tests/mock_network/mod.rs

```diff
@@ -493,228 +501,251 @@ impl SCPNode {
+            let n = i % network_config.nodes.len();
+            simulation.push_value(&node_ids[n], &values[i]);
         }
 
         if last_log.elapsed().as_millis() > 999 {
             log::info!(
-                network.logger,
+                simulation.logger,
                 "( testing ) pushed {}/{} values",
                 i,
-                options.values_to_submit
+                test_options.values_to_submit
             );
             last_log = Instant::now();
         }
 
         let elapsed_duration = Instant::now().duration_since(start);
-        let target_duration = Duration::from_micros(1_000_000 / options.submissions_per_sec);
+        let target_duration = Duration::from_micros(1_000_000 / test_options.submissions_per_sec);
         if let Some(extra_delay) = target_duration.checked_sub(elapsed_duration) {
             std::thread::sleep(extra_delay);
         }
-
-        pushed_values += 1;
     }
 
     // report end of value push
     log::info!(
-        network.logger,
-        "( testing ) pushed {}/{} values",
-        pushed_values,
-        options.values_to_submit
+        simulation.logger,
+        "( testing ) pushed {} values",
+        test_options.values_to_submit
     );
 
     // abort testing if we exceed allowed time
-    let deadline = Instant::now() + options.allowed_test_time;
+    let deadline = Instant::now() + test_options.allowed_test_time;
 
-    // Check that the values got added to the nodes
-    for n in 0..num_nodes as u32 {
-        // Wait for test_node_id(n) to externalize all values
-        let node_id = test_utils::test_node_id(n);
-        let mut prev_num_values = 0;
+    // Check that the values have been externalized by all nodes
+    for node_id in node_ids.iter() {
         let mut last_log = Instant::now();
         loop {
             if Instant::now() > deadline {
                 log::error!(
-                    network.logger,
+                    simulation.logger,
                     "( testing ) failed to externalize all values within {} sec at node {}!",
-                    options.allowed_test_time.as_secs(),
-                    node_id,
+                    test_options.allowed_test_time.as_secs(),
+                    simulation
+                        .names_map
+                        .get(node_id)
+                        .expect("could not find node_id"),
                 );
                 // panic
-                panic!("TEST FAILED DUE TO TIMEOUT");
+                panic!("test failed due to timeout");
             }
 
-            let cur_num_values = network.get_shared_data(&node_id).total_values();
-            if cur_num_values >= values.len() {
+            let num_externalized_values = simulation.get_ledger_size(&node_id);
+            if num_externalized_values >= test_options.values_to_submit {
+                // if the validity_fn does not enforce unique values, we can end up
```

Comment:
Wouldn't it be better to change the validation function to enforce this so that the behavior and test resembles the real world scenario more closely? Maybe this can happen as a preliminary PR?

Reply:
I think you're right and that we should move in this direction. This will require accessing the ledger that we accumulate in `SCPNodeSharedData` in the `validity_fn()`. Probably better for a future PR? I will create a ticket to track.

I think it should be okay to leave this warning in place within `build_and_test()`